### ISSUE TITLE
if_else(): If the Jacobian of any branch evaluates to NaN, the entire Jacobian becomes NaN

### ISSUE THREAD

**Issue by jbaayen** [casadi/casadi#1805]
The following fragment demonstrates the problem:

```
    x = SX.sym('x')
    f = if_else(x > 0, sqrt(x), 0.0, False)
    F = SXFunction([x], [f])
    F.init()
    jac = F.jacobian(0, 0)
    print jac([-1.0])
    print jac([0.0])
    print jac([1.0])
```

This code fragment results in:

```
    -NaN, -NaN, 0.5
```

Whereas I expected to see:

```
     0.0, 0.0, 0.5
```

I assume the culprit is the sqrt(x), which is only defined for x >= 0. The NaN from sqrt(x) is returned, instead of the derivative of the else branch expression.

**Comment by jaeandersson** [comment]
This is actually by design. The non-short-circuiting `if_else` (the only one available for SX) assumes that all branches can be evaluated. You could try replacing `sqrt(x)` with e.g. `abs(sqrt(x))`. (Since the above snippet doesn't correspond to up-to-date CasADi, I can't test for myself)

**Comment by jbaayen** [comment]
Thanks for your quick response Joel. What would be the recommended way to have an MX if_else() where branches can have singularities outside of their respective domains? As I understand, short-circuiting is not yet available for MX if_else() (#1792).

When modelling water systems, we typically have different models for different flow regimes. In some cases, the equations for a particular flow regime contain singularities outside their domain of validity. So being able to 'isolate' branch expressions to their domain of validity is a real must for us.

**Comment by jaeandersson** [comment]
> As I understand, short-circuiting is not yet available for MX if_else() (#1792).

It is available and we use it. But apparently there is a bug present when one branch is trivial (0). If that is blocking you, I'm sure you can work around that bug in CasADi 3.0. I think it should be enough to use to use the more low-level interface: `Function.if_else`.
